segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+        // Increment generation.
+        val previousGen = currentGeneration.getAndIncrement();
+        val oldGen = oldestGeneration.get();
+
+        // Step 1 : Go through all entries once to record counts per each generation.
+        TreeMap<Long, Integer> counts = new TreeMap<Long, Integer>();
+        val iterator = segmentReadIndex.chunkIndex.entrySet().iterator();
+        while (iterator.hasNext()) {
+            val entry = iterator.next();
+            long generation = entry.getValue().getGeneration();
+            val cnt = counts.get(generation);
+            if (null == cnt) {
+                counts.put(generation, 1);
+            } else {
+                counts.put(generation, cnt + 1);
+            }
+        }
+
+        // Step 2 : Determine upto what generation to delete.
+        long deletedUpToGen = 0;
+        int runningCount = 0;
+        for (val entry : counts.entrySet()) {
+            runningCount += entry.getValue();
+            deletedUpToGen = entry.getKey();
+            if (runningCount >= toRemoveCount) {
+                break;
+            }
+        }
+
+        // Step 3: Now remove keys.
+        int removed = 0;
+        val iterator2 = segmentReadIndex.chunkIndex.entrySet().iterator();
+        while (iterator2.hasNext() && removed < toRemoveCount) {
+            val entry = iterator2.next();
+            val gen = entry.getValue().getGeneration();
+            if (gen <= deletedUpToGen) {
+                iterator2.remove();
+                removed++;
+                counts.put(gen, counts.get(gen) - 1);
+            }
+            if (removed == toRemoveCount) {
+                break;
+            }
+        }
+
+        if (removed > 0) {
+            long newOldGenvalue = previousGen + 1;
+            for (val entry : counts.entrySet()) {
+                if (entry.getValue() != 0) {
+                    newOldGenvalue = entry.getKey();
+                    break;
+                }
+            }
+            if (newOldGenvalue != oldGen) {
+                oldestGeneration.compareAndSet(oldGen, newOldGenvalue);
+            }
+            totalChunkCount.getAndAdd(-1 * removed);
+        }
+    }
+
+    /**
+     * Per segment read index.
+     * The index contains mapping from start offset to entries containing chunk names.
+     * The underlying data structure uses ConcurrentSkipListMap.
+     * The generation is used to cleanup unused entries.
+     */
+    @Builder
+    @Data
+    private static class SegmentReadIndex {
+        private final ConcurrentSkipListMap<Long, SegmentReadIndexEntry> chunkIndex;
+        private long generation;
+    }
+
+    /**
+     * Entry for chunk in a segment read index.
+     * The generation is used to cleanup unused entries.
+     */
+    @Builder
+    @Data
+    private static class SegmentReadIndexEntry {
+        private final String chunkName;
+        private long generation;
```

Comment:
This was on purpose.
The left side of the assignment was atomic long. What's the probability of thread getting suspended there and also has been like that for the long time for generation value to matter. 
In worst case, we have early eviction of object. Okay to have that if its rarely , but otherwise  causes overhead.

Well - that was the thinking.
But I fixed it now with atomic anyway.

Reply:
2 threads accessing the same object concurrently is not the only thing you need to worry about in Java. If you set a value to a field, that value may be visible to the current thread, but may not be immediately visible to other threads, even if subsequent threads access the object after the original thread is done with it. You need to use one of the `Atomic*` classes or a mutex (i.e., `synchronized`) to ensure the memory location for that field has been properly set.

Reply:
Tracking this issue here https://github.com/pravega/pravega/issues/4902 This will be done as a separate PR.